### ISSUE TITLE
TOCTOU (time of check vs. time of use) issue in body stream consumption 

### ISSUE THREAD

**Issue by weissi** [vapor/vapor#3271]
https://github.com/vapor/vapor/blob/e1002f35edf92e2a579580f2d1df92e01287c6c7/Sources/Vapor/Request/Request%2BBody.swift#L43-L46

The linked code has an issue. It's something like

```
    switch self.request.bodyStorage.withLockedValue({ $0 }) { // LOCK ACQUSITION
    case .stream(let stream): // ACT UPON value
        return stream.consume(max: max, on: self.request.eventLoop).map { buffer in
            self.request.bodyStorage.withLockedValue({ $0 = .collected(buffer) }) // LOCK RE-ACQUISITON & OVERWRITE
```

Any code that re-acquires locks and assumes that the state underneath hasn't changed isn't quite right.

If `collect()` is intended to only be run once, this should be documented. And to catch users doing a bad thing, it should probably be changed to something like

```
switch self.request.bodyStorage.withLockedValue({ value in let oldValue = value; value = .currentlyStreaming;  return oldValue })
```

and then you can throw an error if the user attempts to consume the body twice whilst it's streaming.

Most importantly, we should document exactly how `collect()` is supposed to be used.

**Comment by gwynne** [comment]
IIRC this was assumed to be "safe" (very heavy on the quotes here) because there would never be a modification of the state off the event loop. That doesn't make this code any less incorrect, to be sure.

**Comment by weissi** [comment]
> IIRC this was assumed to be "safe" (very heavy on the quotes here) because there would never be a modification of the state off the event loop. That doesn't make this code any less incorrect, to be sure.

Right, from a low-level data race point of view, this is safe. Even if it were called on another thread as the access is thread.

We have a high-level data race here (e.g. something neither the Swift Compiler nor TSan can catch) which is a logic issue. Even in one EventLoop we could call `collect()` twice. E.g.

```
{ request in
    let yo = request.body.collect()
    let lo = request.body.collect()

    return yo.and(lo).whenComplete { print("whoops", yo, lo) }
}
```

**Comment by weissi** [comment]
The [docs](https://swiftinit.org/docs/vapor/vapor/httpbodystreamstrategy.collect) don't specify what the behaviour for multi-collect is supposed to be.

**Comment by weissi** [comment]
FWIW, the real world problem we hit looked like this

```swift
func collectBody1(_ req: Request) async throws -> Foo {
  let collectedBody: ByteBuffer = try await req.body.collect(max: maxBodySize).get()
  // Do something with body
}

func collectBody2(_ req: Request) async throws -> Bar {
  let collectedBody: ByteBuffer = try await req.body.collect(max: maxBodySize).get()
  // Do something else with body
}

async let a = collectBody1(req)
async let b = collectBody2(req)

await a
await b
// collectBody1 and 2 each get garbled subsets of the streaming body
```

So like the `yo.and(lo)` example but probably easier to follow

**Comment by gwynne** [comment]
I don't think `collect()` is supposed to be called multiple times for the same request, but I'm not 100% sure. @0xTim?

**Comment by weissi** [comment]
> I don't think `collect()` is supposed to be called multiple times for the same request, but I'm not 100% sure. @0xTim?

At first, I thought that might be the case, but then you see this code:

```
return stream.consume(max: max, on: self.request.eventLoop).map { buffer in
    self.request.bodyStorage.withLockedValue({ $0 = .collected(buffer) })
```

So once it has _finished_ streaming in the body, it does change it to `.collected(buffer)`. So
- it's paying the memory cost to keep it in memory for the remaining lifetime of the request
- if you collect it 'late enough' it _will_ work
- so currently it's just racy and that's surprising to a user

So if you were to change it to `self.request.bodyStorage.withLockedValue({ $0 = .alreadyCollectedSorry })` that would break some existing users. Therefore, the only thing that makes sense to me is to just fix it. It should be as simple as introducing a `.collecting(Future)` state and populating it with the future that's running the original collection.

**Comment by 0xTim** [comment]
Yeah I don't think calling it multiple times was intended to be supported, it just works in some cases. I'm hesitant to go messing around with the state machine given how flaky it is, but will take a look
